### PR TITLE
[AssetMapper] Check asset/vendor directory is writable

### LINKED ISSUE #54719: [AssetMapper] importmap:install silently fail but say everything is ok on prod
### Symfony version(s) affected

6.4.6

### Description

If you have no permissions to write to `assets` directrory you'll just get a php warning in prod enviroment (as warnings does not converts to exceptions there).
It is not right to say "all right" when it is not I think. And return exit code 0. bash scripts with `set -e` continue executing with not installed assets in CI (for example). Good that `assets:compile` throw exeption like "you did not run importmap:install", but problem was in another place.

```
21:14:28 WARNING   [php] Warning: file_put_contents(/var/www/html/assets/vendor/installed.php): Failed to open stream: No such file or directory ["exception" => ErrorException { …}]

                                                                                                                        
 [OK] Downloaded 1 package into assets/vendor.                                                                          
                                                                                                                        

```

### How to reproduce

```bash
./bin/console imp:req bootstrap
rm -rf assets/vendor
echo 'APP_ENV=prod' >> .env.local
chmod 555 assets
./bin/console imp:inst
```

### Possible Solution

Check that downloded package exists and throw exception otherwise.

### PR DESCRIPTION
**Pull request by smnandre**
| Q             | A
| ------------- | ---
| Branch?       | 6.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Issues        | Fix #54719
| License       | MIT

💐

### CODE REVIEW

**Comment by smnandre** [comment]
Failures unrelated (Predis nullables)

File: src/Symfony/Component/AssetMapper/ImportMap/RemotePackageStorage.php [hunk 1]
```diff
@@ -90,6 +88,18 @@ public function getDownloadPath(string $packageModuleSpecifier, ImportMapType $i
...
+    private function saveFile(string $filename, string $contents): void
+    {
+        if (!is_dir($vendorDir = \dirname($filename)) && !@mkdir($vendorDir, 0777, true)) {
+            throw new RuntimeException(sprintf('The directory "%s" does not exist and could not be created.', $vendorDir));
+        }
+        if (!is_writable($vendorDir)) {
+            throw new RuntimeException(sprintf('The directory "%s" is not writable.', $vendorDir));
+        }
+
+        file_put_contents($filename, $contents);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/AssetMapper/ImportMap/RemotePackageStorage.php]: if (false === etc then throw?

Comment (smnandre) [reply src/Symfony/Component/AssetMapper/ImportMap/RemotePackageStorage.php]: If you meant: 
```php
        if (false === file_put_contents($filename, $contents)) {
            throw new RuntimeException(sprintf('The file "%s" could not be written to.', $filename));
        }
```        

Updated, thanks :)

**Review by nicolas-grekas — APPROVED** [review]

File: src/Symfony/Component/AssetMapper/Tests/ImportMap/RemotePackageStorageTest.php [hunk 2]
```diff
@@ -41,6 +41,21 @@ public function testGetStorageDir()
         $this->assertSame(realpath(self::$writableRoot.'/assets/vendor'), realpath($storage->getStorageDir()));
     }
 
+    public function testSaveThrowsWhenVendorDirectoryIsNotWritable()
+    {
+        if (!file_exists($vendorDir = self::$writableRoot.'/assets/acme/vendor')) {
```
Comment (OskarStark) [inline src/Symfony/Component/AssetMapper/Tests/ImportMap/RemotePackageStorageTest.php]: IIRC there is a new method on filesystem for that

File: src/Symfony/Component/AssetMapper/ImportMap/RemotePackageStorage.php [hunk 3]
```diff
@@ -52,7 +54,9 @@ public function save(ImportMapEntry $entry, string $contents): void
         $vendorPath = $this->getDownloadPath($entry->packageModuleSpecifier, $entry->type);
 
         @mkdir(\dirname($vendorPath), 0777, true);
```
Comment (OskarStark) [inline src/Symfony/Component/AssetMapper/ImportMap/RemotePackageStorage.php:56]: Why not using Symfony FS here, it is already a dependency, isn't it?

File: src/Symfony/Component/AssetMapper/Tests/ImportMap/RemotePackageStorageTest.php (hunk 2 above)
Comment (smnandre) [reply src/Symfony/Component/AssetMapper/Tests/ImportMap/RemotePackageStorageTest.php]: On 6.4 too ?

File: src/Symfony/Component/AssetMapper/ImportMap/RemotePackageStorage.php (hunk 3 above)
Comment (smnandre) [reply src/Symfony/Component/AssetMapper/ImportMap/RemotePackageStorage.php:56]: I did not want to refactor only partially this class... but i noted to myself to come back later :)

File: src/Symfony/Component/AssetMapper/Tests/ImportMap/RemotePackageStorageTest.php (hunk 2 above)
Comment (smnandre) [reply src/Symfony/Component/AssetMapper/Tests/ImportMap/RemotePackageStorageTest.php]: Updated the tests

File: src/Symfony/Component/AssetMapper/ImportMap/RemotePackageStorage.php [hunk 4]
```diff
@@ -52,7 +54,9 @@ public function save(ImportMapEntry $entry, string $contents): void
...
-        file_put_contents($vendorPath, $contents);
+        if (false === @file_put_contents($vendorPath, $contents)) {
+            throw new RuntimeException(error_get_last()['message'] ?? sprintf('The file "%s" could not be written to.', $vendorPath));
```
Comment (rosier) [inline src/Symfony/Component/AssetMapper/ImportMap/RemotePackageStorage.php]:
```suggestion
            throw new RuntimeException(error_get_last()['message'] ?? sprintf('Failed to write file "%s".', $vendorPath));
```

File: src/Symfony/Component/AssetMapper/ImportMap/RemotePackageStorage.php [hunk 5]
```diff
@@ -64,7 +68,9 @@ public function saveExtraFile(ImportMapEntry $entry, string $extraFilename, stri
         $vendorPath = $this->getExtraFileDownloadPath($entry, $extraFilename);
 
         @mkdir(\dirname($vendorPath), 0777, true);
-        file_put_contents($vendorPath, $contents);
+        if (false === @file_put_contents($vendorPath, $contents)) {
+            throw new RuntimeException(error_get_last()['message'] ?? sprintf('The file "%s" could not be written to.', $vendorPath));
```
Comment (rosier) [inline src/Symfony/Component/AssetMapper/ImportMap/RemotePackageStorage.php]: Same here

Comment (smnandre) [inline src/Symfony/Component/AssetMapper/ImportMap/RemotePackageStorage.php]:
```suggestion
            throw new RuntimeException(error_get_last()['message'] ?? sprintf('Failed to write file "%s".', $vendorPath));
```

Comment (smnandre) [reply src/Symfony/Component/AssetMapper/ImportMap/RemotePackageStorage.php]: Thanks

**Review by rosier — APPROVED** [review]

**Review by OskarStark — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @smnandre.

**Pull request merged**